File: 프로그래머스/0/181881. 조건에 맞게 수열 변환하기 2/조건에 맞게 수열 변환하기 2.py
import copy
def solution(arr):
    answer = 0
    cnt = 0
    
    while True:
        old = copy.deepcopy(arr)
        for k, a in enumerate(arr):
            if a >= 50 and a % 2 == 0:
                arr[k] = a // 2
            elif a < 50 and a % 2 == 1:
                arr[k] = a * 2 +1
        if old == arr:
            break
        cnt += 1
        
    answer = cnt
    return answer

"""
arr 원소: 50보다 크거나 같고 짝수면 /2
50보다 작고 홀수면 *2+1
이때 arr(x) == arr(x+1)일 때의 x를 구하기


"""

File: 프로그래머스/0/181881. 조건에 맞게 수열 변환하기 2/조건에 맞게 수열 변환하기 2.java
class Solution {
    public int solution(int[] arr) {
        int answer = 0;
        int[] preview = new int[arr.length];
        int[] after = new int[arr.length];
        
        for (int i = 0; i < arr.length; i++) {
            preview[i] = arr[i];
        }
        
        int cnt = 0;
        
        while(true) {
            boolean b = true;
            for (int i = 0; i < after.length; i++) {
                if (preview[i] >= 50 && preview[i] % 2 == 0) {
                    after[i] = preview[i] / 2;
                } else if (preview[i] < 50 && preview[i] % 2 == 1) {
                    after[i] = preview[i] * 2 + 1;
                } else {
                    after[i] = preview[i];
                }
                           
                if (after[i] != preview[i]) {
                    b = false;
                }
            }
                           
            if (b) {
                answer = cnt;
                break;
            } else {
                for (int i = 0; i < after.length; i++) {
                    preview[i] = after[i];
                }
            }
            cnt++;
        }
        
        return answer;
    }
}